### ISSUE TITLE
GIF Regression? Error: Input file has corrupt header: gifload: Failed to read from given file

### ISSUE THREAD

**Issue by mooyoul** [lovell/sharp#1890]
### What is the output of running `npx envinfo --binaries --languages --system --utilities`?

```
  System:
    OS: macOS Mojave 10.14.6
    CPU: (8) x64 Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
    Memory: 975.51 MB / 16.00 GB
    Shell: 3.0.2 - /usr/local/bin/fish
  Binaries:
    Node: 8.14.0 - /usr/local/bin/node
    Yarn: 1.15.2 - /usr/local/bin/yarn
    npm: 6.11.3 - /usr/local/bin/npm
  Utilities:
    CMake: 3.14.0 - /opt/local/bin/cmake
    Make: 3.81 - /usr/bin/make
    GCC: 10.14. - /usr/bin/gcc
    Git: 2.20.1 - /usr/bin/git
    Clang: 1001.0.46.4 - /usr/bin/clang
    Mercurial: 2.8.1+20131205 - /usr/local/bin/hg
    Subversion: 1.10.3 - /usr/bin/svn
    FFmpeg: 4.1.1 - /opt/local/bin/ffmpeg
  Languages:
    Bash: 3.2.57 - /bin/bash
    Go: 1.11.4 - /usr/local/go/bin/go
    Java: 1.8.0 - /usr/bin/javac
    Perl: 5.18.4 - /usr/bin/perl
    PHP: 7.1.23 - /usr/bin/php
    Python: 2.7.10 - /usr/bin/python
    Ruby: 2.4.2 - /Users/Prescott/.rbenv/shims/ruby
    Rust: 1.13.0 - /Users/Prescott/.cargo/bin/rustup
    Scala: 2.12.2 - /usr/local/scala/bin/scalac
```

### What are the steps to reproduce?

1. Install sharp@0.23.0 or sharp@0.23.1
2. Download [sample image](https://test-mooyoul.s3-accelerate.amazonaws.com/public/sharp-gif-regression/input.gif) as `input.gif`
3. Save below code to `test.js` and run `node test.js`

```js
'use strict';
const fs = require('fs');
const sharp = require('sharp');
const sharpPkg = require('sharp/package.json');
const os = require('os');

const FIXTURE_PATH = 'input.gif';

const suites = [{
  label: 'with sharp(filename)',
  test: async () => {
    await sharp(FIXTURE_PATH, { failOnError: false }).metadata();
  },
}, {
  label: 'with sharp(buffer)',
  test: async () => {
    await sharp(fs.readFileSync(FIXTURE_PATH, { failOnError: false })).png().toBuffer();
  },
}];

(async () => {
  console.log('%s %s', os.platform(), os.arch());
  console.log('using sharp %s. sharp.versions: ', sharpPkg.version, sharp.versions);

  for (const suite of suites) {
    console.log('Trying %s', suite.label);

    try {
      await suite.test();
      console.log('PASSED %s', suite.label);
    } catch (e) {
      console.error('FAILED %s', suite.label);
      console.error(e);
    }
  }
})().catch(console.error);
```

### What is the expected behaviour?

Transformation should work.

### Are you able to provide a standalone code sample, without other dependencies, that demonstrates this problem?

Yes, See upper section. Also Screencast is available: https://asciinema.org/a/AZ7Z3wDn2Ro8oEcwq1pZE0gPe

### Are you able to provide a sample image that helps explain the problem?

Yes. https://test-mooyoul.s3-accelerate.amazonaws.com/public/sharp-gif-regression/input.gif

##### for issue search results

Error: Input file has corrupt header: gifload: Failed to read from given file
Error: Input buffer has corrupt header: gifload_buffer: Failed to read from given file

**Comment by mooyoul** [comment]
Oh, I figured out why it's failing. The image is really corrupted 🤦‍♂
According to gifsicle, Attached GIF image has 42 frames, but last frame has insufficient image data.

<img width="535" alt="스크린샷 2019-09-27 오후 7 40 20" src="https://user-images.githubusercontent.com/2101743/65763847-3b077d00-e15f-11e9-96c6-dd4c106a2954.png">

If i extract first frame using gifsicle and feed it to sharp. It works.
<img width="1275" alt="스크린샷 2019-09-27 오후 7 46 26" src="https://user-images.githubusercontent.com/2101743/65764011-96396f80-e15f-11e9-99e2-c1bb41a42f99.png">

Anyway, Now i'm curious why recent sharp build fails with this corrupted gif image.
1) Does sharp try to decode entire frames of Animated GIF? If so, why?
2) Is there any option to use "first frame only"? like `failOnError` option do

**Comment by lovell** [comment]
Are you using the prebuilt libvips binaries? sharp v0.22.x provides libvips v8.7.4 whilst sharp v0.23.x provides libvips v8.8.1 - see https://sharp.pixelplumbing.com/en/stable/changelog/

The libvips changelog is at https://github.com/libvips/libvips/blob/master/ChangeLog - you'll see there were improvements to gifload to improve strictness and memory usage, although some were a bit too strict and I think you might have run into the problem that https://github.com/libvips/libvips/commit/25e457736173369dcb0f7c09d07af68aedbdc175 addresses.

This was released as part of libvips v8.8.2, so the future sharp v0.24.x will provide a prebuilt libvips with this fix.

**Comment by mooyoul** [comment]
Oh I see. Thanks for your kind explaination 👍 then I'll wait for next release.

Also, This is my temporal workaround to resolve this issue. Hope this helps!

```typescript
import * as execa from "execa";
import * as sharp from "sharp";

const BUF_GIF89A = Buffer.from([0x47, 0x49, 0x46, 0x38, 0x39, 0x61]); // "GIF" file signature + 89a extension

class ImageService {
  public async transform(buf: Buffer): Promise<Buffer> {
    return await this.load(buf).resize(200, 200).jpg().toBuffer();
  }

  public async load(buf: Buffer): Promise<sharp.Sharp> {
    const isAnimatedGIF = Buffer.compare(buf.slice(0, BUF_GIF89A.length), BUF_GIF89A) === 0;

    // if given input is animated GIF, extract first frame and feed to sharp
    if (isAnimatedGIF) {
       const bufFrame = await this.extractFrame(buf);
       return sharp(bufFrame, { failOnError: false });
    }

    return sharp(buf, { failOnError: false });
  }

  public async extractFrame(buf: Buffer): Promise<Buffer> {
    const res = await execa("gifsicle", ["-", "#0"], { input: buf, encoding: null });

    return res.stdout;
  }
}
```

**Comment by mooyoul** [comment]
BTW, Is there any technical details described about https://github.com/lovell/sharp/issues/1578#issuecomment-464311175 ? Just curious.

**Comment by lovell** [comment]
@mooyoul I think the Samsung firmware is misusing JPEG markers for some kind of metadata but I've not looked into it. There's a lot of evidence of the problem out on the web e.g. https://www.google.com/search?q=%22Invalid+SOS+parameters%22+samsung

**Comment by mooyoul** [comment]
Oh, Thanks for the keyword. I owe you so many things 🙏I really like sharp. Thank you for making this cool software!
- with love from Seoul

**Comment by lovell** [comment]
sharp v0.24.0 now available with libvips v8.9.0
